### PR TITLE
Remove module tabs

### LINKED ISSUE #159: Indicate whether a layer is disabled on the icon
Icons for LayerTabs should indicate if the layer is currently disabled (e.g. red cross over icon).

### LINKED ISSUE #231: Don't allow modules in separate tabs
**Describe the feature / issue**
At present, double-clicking a module in a layer opens up the module in a separate tab. This can eventually create a considerable list of tabs in the main window.  There is enough space for the additional module controls / graphs to be displayed alongside the keyword options.

**Additional details**
- Configuration targets list should be moved to be adjacent to the Module header information.
- Controls for all widgets should be placed in tabs for consistency.

### PR DESCRIPTION
**Pull request by trisyoungs**
This PR changes the way module controls are accessed in the GUI. Existing behaviour was for only module keywords to be displayed within the parent LayerTab - if access to additional controls (principally graphing) was required then the module had to be double-clicked to open up a new tab.

The present PR removes module tabs completely, preferring to display keywords and additional controls for a module in separate tabs within the LayerTab, reducing context switching. It also prevents tab overflow on smaller screens when many tabs are opened.

Tutorials need to be updated to reflect these changes, but those changes will be made along with others enforced by a separate PR from the `sq-averaging` branch.

Closes #231.
Closes #159.

### CODE REVIEW

File: src/gui/modulecontrolwidget.h [hunk 1]
```diff
@@ -21,10 +21,13 @@
 
 #pragma once
 
+#include "base/lock.h"
```
Comment (DanNixon) [inline src/gui/modulecontrolwidget.h:24]: `Lock` is pretty much a `std::counting_semaphore` (and I cannot really tell what purpose `Locker` serves).

Separate to this, but we should look at replacing those. There are also a lot nicer synchronisation methods in the thread support section of the standard library.

**Review by DanNixon — COMMENTED** [review]
Looks good to me, I'll let @rprospero take a closer look since he is likely more familiar with the GUI than me.

As a side note, have you considered using a standard icon pack (e.g. Font Awesome or MDI) rather than creating custom ones for pretty standard features?

File: src/gui/modulecontrolwidget_funcs.cpp [hunk 2]
```diff
@@ -69,16 +71,30 @@ void ModuleControlWidget::setModule(Module *module, Dissolve *dissolve)
...
+                                                        QString::fromStdString(std::string(module->type()))));
+    ui_.ModuleIconLabel->setPixmap(ModuleBlock::modulePixmap(module_));
 
-    // Set up our keywords widget
+    // Set up our control widgets
     ui_.ModuleKeywordsWidget->setUp(module_->keywords(), dissolve_->constCoreData());
 
+    // Create any additional controls offered by the Module
+    moduleWidget_ = module->createWidget(NULL, *dissolve_);
+    if (moduleWidget_ == NULL)
```
Comment (rprospero) [inline src/gui/modulecontrolwidget_funcs.cpp:84]: nullptr?

File: src/gui/modulelisteditor_funcs.cpp [hunk 3]
```diff
@@ -76,11 +77,11 @@ bool ModuleListEditor::setUp(DissolveWindow *dissolveWindow, ModuleLayer *module
...
         RefDataListIterator<MimeTreeWidgetItem, QString> categoryIterator(moduleCategories_);
         while ((categoryItem = categoryIterator.iterate()))
-            if (categoryIterator.currentData() == QString::fromStdString(std::string(module->category())))
+            if (DissolveSys::sameString(module->category(), qPrintable(categoryIterator.currentData())))
                 break;
-        if (categoryItem == nullptr)
+        if (categoryItem == NULL)
         {
-            categoryItem = new MimeTreeWidgetItem((QTreeWidget *)nullptr, 1000);
+            categoryItem = new MimeTreeWidgetItem((QTreeWidget *)NULL, 1000);
```
Comment (rprospero) [inline src/gui/modulelisteditor_funcs.cpp:84]: nullptr?

File: src/gui/modulelisteditor_funcs.cpp [hunk 4]
```diff
@@ -146,15 +143,25 @@ void ModuleListEditor::disableSensitiveControls()
 {
     ui_.AvailableModulesTree->setEnabled(false);
     chartWidget_->disableSensitiveControls();
-    ui_.ControlsWidget->disableSensitiveControls();
+    for (int n = 0; n < ui_.ModuleControlsStack->count(); ++n)
+    {
+        ModuleControlWidget *mcw = dynamic_cast<ModuleControlWidget *>(ui_.ModuleControlsStack->widget(n));
+        if (mcw)
+            mcw->disableSensitiveControls();
+    }
```
Comment (rprospero) [inline src/gui/modulelisteditor_funcs.cpp:151]:
```suggestion
    for (ModuleControlWidget *mcw : ui_.ModuleControlsStack)
    {
        if (mcw)
            mcw->disableSensitiveControls();
    }
```

Same caveat as above

File: src/gui/modulelisteditor_funcs.cpp [hunk 5]
```diff
@@ -146,15 +143,25 @@ void ModuleListEditor::disableSensitiveControls()
...
 {
     ui_.AvailableModulesTree->setEnabled(true);
     chartWidget_->enableSensitiveControls();
-    ui_.ControlsWidget->enableSensitiveControls();
+    for (int n = 0; n < ui_.ModuleControlsStack->count(); ++n)
+    {
+        ModuleControlWidget *mcw = dynamic_cast<ModuleControlWidget *>(ui_.ModuleControlsStack->widget(n));
+        if (mcw)
+            mcw->enableSensitiveControls();
+    }
```
Comment (rprospero) [inline src/gui/modulelisteditor_funcs.cpp:164]:
```suggestion
    for (ModuleControlWidget *mcw : ui_.ModuleControlsStack)
    {
        if (mcw)
            mcw->enableSensitiveControls();
    }
```

File: src/gui/modulelisteditor_funcs.cpp [hunk 6]
```diff
@@ -164,26 +171,80 @@ void ModuleListEditor::setModulePaletteVisible(bool visible) { ui_.ModulePalette
...
-void ModuleListEditor::blockSelectionChanged(const QString &blockIdentifier)
+// Find the ModuleControlWidget for this Module in the stack, if it exists
+ModuleControlWidget *ModuleListEditor::controlWidgetForModule(Module *module) const
+{
+    for (int n = 0; n < ui_.ModuleControlsStack->count(); ++n)
+    {
+        ModuleControlWidget *mcw = dynamic_cast<ModuleControlWidget *>(ui_.ModuleControlsStack->widget(n));
+        if (mcw && (mcw->module() == module))
+            return mcw;
+    }
```
Comment (rprospero) [inline src/gui/modulelisteditor_funcs.cpp:182]:
```suggestion
    for (ModuleControlWidget *mcw : ui_.ModuleControlsStack)
    {
        if (mcw && (mcw->module() == module))
            return mcw;
    }
```

File: src/gui/modulelisteditor_funcs.cpp [hunk 7]
```diff
@@ -164,26 +171,80 @@ void ModuleListEditor::setModulePaletteVisible(bool visible) { ui_.ModulePalette
...
+
+// Return the index of the ModuleControlWidget for this Module in the stack, if it exists
+int ModuleListEditor::widgetIndexForModule(Module *module) const
+{
+    for (int n = 0; n < ui_.ModuleControlsStack->count(); ++n)
+    {
+        ModuleControlWidget *mcw = dynamic_cast<ModuleControlWidget *>(ui_.ModuleControlsStack->widget(n));
+        if (mcw && (mcw->module() == module))
+            return n;
+    }
```
Comment (rprospero) [inline src/gui/modulelisteditor_funcs.cpp:195]:
```suggestion
    for (ModuleControlWidget *mcw : ui_.ModuleControlsStack)
    {
        if (mcw && (mcw->module() == module))
            return n;
    }
```

File: src/keywords/configurationreflist.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+/*
+ * Arguments
+ */
+
+// Return minimum number of arguments accepted
+int ConfigurationRefListKeyword::minArguments() const { return 1; }
+
+// Return maximum number of arguments accepted
+int ConfigurationRefListKeyword::maxArguments() const { return 99; }
```
Comment (rprospero) [inline src/keywords/configurationreflist.cpp:55]: Maybe MAX_INT?  It's overkill, but then so is 99

**Review by rprospero — COMMENTED** [review]
Mostly just tightening up some for loops.  Be forewarned that I haven't actually tested these patches.  In fact, I'm pretty sure that two of them are mutually exclusive, so accepting them all will break the build.

File: src/gui/modulecontrolwidget.h (hunk 1 above)
Comment (trisyoungs) [reply src/gui/modulecontrolwidget.h:24]: Noted as #395.

File: src/gui/modulelisteditor.h [hunk 9]
```diff
@@ -79,7 +82,7 @@ class ModuleListEditor : public QWidget
 
     public:
     // Set up the ModuleListEditor for the specified ModuleLayer
-    bool setUp(DissolveWindow *dissolveWindow, ModuleLayer *moduleLayer, Configuration *localConfiguration = nullptr);
+    bool setUp(DissolveWindow *dissolveWindow, ModuleLayer *moduleLayer, Configuration *localConfiguration = NULL);
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/gui/modulelisteditor.h:85]: Bad merging!

File: src/gui/keywordwidgets/configurationreflist_funcs.cpp [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    if (refreshing_)
+        return;
+
+    refreshing_ = true;
+
+    // Ensure that we obey any limit on maximum number of selected items
+    auto nChecked = 0;
+    for (int n = 0; n < ui_.SelectionList->count(); ++n)
+        if (ui_.SelectionList->item(n)->checkState() == Qt::Checked)
+            ++nChecked;
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/gui/keywordwidgets/configurationreflist_funcs.cpp:89]: Good thought, but unfortunately Qt doesn't make things that easy.  There is a `QListWidget::items()` function returning a `QList` which *is* iterable, but the `items()` function looks to be for use with `QMimeData`, and so is unhelpful.

So, consider yourself mildly surprised!

File: src/gui/modulelisteditor_funcs.cpp (hunk 4 above)
Comment (trisyoungs) [reply src/gui/modulelisteditor_funcs.cpp:151]: Alas, also a "No". `QStackedWidget` does have a `children()` function that returns an iterable `QObjectList`, but attempting to cast these up to the necessary `ModuleControlWidget`s gives a hard crash.

File: src/keywords/configurationreflist.cpp (hunk 8 above)
Comment (trisyoungs) [reply src/keywords/configurationreflist.cpp:55]: Probably a good place for a `std::optional<int>`, returning `nullopt` if there is no set limit. Added as issue #403.

**Pull request merged**